### PR TITLE
Add redis engine for offline store

### PR DESCRIPTION
**Pull request by dhruvjain99**
## Proposed Changes

Add a new offline message store engine - Redis

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #XXXX)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CODE_OF_CONDUCT.md` document
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if needed)
- [ ] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [ ] I have squashed all my commits into one before merging

### CODE REVIEW

File: apps/vmq_generic_offline_msg_store/priv/vmq_generic_offline_msg_store.schema [hunk 1]
```diff
@@ -12,26 +12,26 @@
...
 {mapping, "offline_message_store_opts.password", "vmq_generic_offline_msg_store.msg_store_opts.password",
-    [{default, "vmq_test_password"},
-     {datatype, string}
+    [{default, undefined},
+     {datatype, atom}
     ]}.
 
 {mapping, "offline_message_store_opts.database", "vmq_generic_offline_msg_store.msg_store_opts.database",
-    [{default, "vmq_test_database"},
+    [{default, "2"},
```
Comment (dumbpreacher) [inline apps/vmq_generic_offline_msg_store/priv/vmq_generic_offline_msg_store.schema:30]: Why is this set to 2?

File: apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    undefined -> undefined;
+                    User when is_atom(User) -> atom_to_list(User)
+                end,
+    Password = case proplists:get_value(password, Opts, undefined) of
+                    undefined -> undefined;
+                    Pass when is_atom(Pass) -> atom_to_list(Pass)
+                end,
+    {Database, _} = string:to_integer(proplists:get_value(database, Opts, "2")),
+    Port = proplists:get_value(port, Opts, 26379),
+    SentinelHosts = vmq_schema_util:parse_list(proplists:get_value(host, Opts, "[\"localhost\"]")),
```
Comment (dumbpreacher) [inline apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl:21]: Was it decided that offline store needs to be a Sentinel?

File: apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            open_(Opts)
+    end.
+
+write(Client, SIdB, _MsgRef, MsgB, Timeout) ->
+    vmq_redis:query(Client, ["RPUSH", SIdB, MsgB], ?RPUSH, ?MSG_STORE_WRITE, Timeout).
+
+delete(Client, SIdB, Timeout) ->
+    vmq_redis:query(Client, ["DEL", SIdB], ?DEL, ?MSG_STORE_DELETE, Timeout).
+
+delete(Client, SIdB, _MsgRef, Timeout) ->
```
Comment (dumbpreacher) [inline apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl:45]: The MsgRef arg is unnecessary, isn't it?

File: apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    vmq_redis:query(Client, ["LPOP", SIdB, 1], ?LPOP, ?MSG_STORE_DELETE, Timeout).
+
+read(_Client, _SIdB, _MsgRef, _Timeout) ->
+    {error, not_supported}.
+
+find(Client, SIdB, Timeout) ->
```
Comment (dumbpreacher) [inline apps/vmq_generic_offline_msg_store/src/engines/vmq_offline_storage_engine_redis.erl:51]: Where is this function being used?

File: apps/vmq_server/src/vmq_message_store.erl [hunk 5]
```diff
@@ -77,6 +76,7 @@ read_with_retry(SubscriberId, MsgRef, N) when N >= 0 ->
     case vmq_plugin:only(msg_store_read, [SubscriberId, MsgRef]) of
         {ok, _} = OkRes -> OkRes;
         {error, no_matching_hook_found} = ErrRes -> ErrRes;
+        {error, not_supported} = ErrRes -> ErrRes;
```
Comment (dumbpreacher) [inline apps/vmq_server/src/vmq_message_store.erl:79]: not_supported and no_matching_hook_found can be handled in one case.

File: apps/vmq_server/test/vmq_info_SUITE.erl [hunk 6]
```diff
@@ -168,21 +167,21 @@ filtering_works(_Config) ->
...
-                 "--client_id", "--routing_key"]),
-
-    [#{client_id := <<"vmq-info-client">>, dup := false}] =
-        execute(["vmq-admin", "session", "show", "--dup=false",
-                 "--client_id", "--dup"]),
-
-    [#{client_id := <<"vmq-info-client">>, payload := <<"thepayload">>}] =
-        execute(["vmq-admin", "session", "show", "--payload=thepayload",
-                 "--client_id", "--payload"]),
+%%    [#{client_id := <<"vmq-info-client">>, msg_qos := 1}] =
```
Comment (dumbpreacher) [inline apps/vmq_server/test/vmq_info_SUITE.erl:170]: Clean up commented code?

Comment (dhruvjain99) [reply apps/vmq_server/test/vmq_info_SUITE.erl]: The commented tests are fixed.

**Review by dumbpreacher — APPROVED** [review]

**Pull request merged**